lib/msal-browser/src/broker/client/BrokerClientApplication.ts

```diff
@@ -30,24 +27,44 @@ import { BrokerSsoSilentRequest } from "../request/BrokerSsoSilentRequest";
  */
 export class BrokerClientApplication extends ClientApplication {
 
-    private cachedBrokerResponse: Promise<BrokerAuthenticationResult>;
+    private currentBrokerRedirectResponse: Promise<BrokerAuthenticationResult>;
 
     constructor(configuration: Configuration) {
         super(configuration);
     }
 
     /**
-     * 
+     * Event handler function which allows users to fire events after the PublicClientApplication object
+     * has loaded during redirect flows. This should be invoked on all page loads involved in redirect
+     * auth flows.
+     * @param hash Hash to process. Defaults to the current value of window.location.hash. Only needs to be provided explicitly if the response to be handled is not contained in the current value.
+     * @returns {Promise.<AuthenticationResult | null>} token response or null. If the return value is null, then no auth redirect was detected.
+     */
+    async handleRedirectPromise(hash?: string): Promise<BrokerAuthenticationResult | null> {
+        const brokerResponse = super.handleRedirectPromise(hash) as Promise<BrokerAuthenticationResult>;
+        this.currentBrokerRedirectResponse = this.waitForBrokeredResponse(brokerResponse);
+        const redirectResponse = await this.currentBrokerRedirectResponse;
+        this.currentBrokerRedirectResponse = undefined;
```

Comment:
This feels a little hacky but I can't think of an alternative at the moment. This may be something we should do in the base `handleRedirectPromise` to prevent the race conditions we hear about from time to time.

Reply:
It is a bit hacky, I needed a way to wait for the response to finish in case the broker receives a request for the redirect response before it is done trading for the tokens. Happy to hear alternatives. The reason I haven't done this in the base `handleRedirectPromise()` is because I am trying to keep broker specific code away from the base functions.

Reply:
I'm advocating for making this not broker specific. We have scenarios where if you call `handleRedirectPromise` twice at the same time it has the potential to create a race condition and break. I think moving this logic up one level, into the base function would solve both problems. Ultimately what you're trying to prevent here is having more than one instance of `handleRedirectPromise` running concurrently, right?

Reply:
That is more of a side effect, the main reason is that I need a way to determine if we are currently trading for tokens or not, so that when an embedded app makes a request we can wait for the tokens to be retrieved before returning a response back from the broker.